File: app/lib/rinex_merge.php
<?php

declare(strict_types=1);

/**
 * Склейка нескольких часовых RINEX-файлов одной станции/типа в файлы
 * "за сутки" (см. bin/process_rinex_requests.php). Правила:
 *   1) Склеиваем только файлы БЕЗ разрыва между ними — если конец одного
 *      файла (его начало + номинальная длительность из имени) и начало
 *      следующего расходятся больше чем на $gapToleranceMinutes, это два
 *      разных непрерывных отрезка, и из них получаются ДВА отдельных
 *      файла, а не один с дыркой посередине.
 *   2) Склейка никогда не пересекает границу суток (UTC) — даже идущие
 *      впритык файлы 23:00 и 00:00 следующего дня попадут в разные
 *      сутки и не объединятся, по тому же принципу, по которому организован
 *      сам каталог на FTP (один день — одна папка).
 *
 * Объединение самого RINEX-содержимого — построчное: у файла одна шапка до
 * строки "END OF HEADER" (включительно), дальше данные. Для склеенного
 * файла шапка берётся из первого (по времени) файла группы, у всех
 * остальных шапка вырезается целиком, остаются только данные. Работает
 * одинаково для observation- и navigation-файлов — оба формата размечают
 * конец шапки той же меткой "END OF HEADER".
 */

/**
 * Группирует список файлов (каждый — ['local_path', 'timestamp' (DateTimeImmutable),
 * 'period_minutes' (?int)]) на непрерывные отрезки. Файлы должны принадлежать
 * одной станции и одному типу (MO либо MN) — группировка по станции/типу/дню
 * делается ДО вызова этой функции, на уровне bin/process_rinex_requests.php.
 *
 * @return array<int, array<int, array>> список групп, каждая — список файлов в хронологическом порядке
 */
function rinex_group_contiguous_files(array $files, int $gapToleranceMinutes = 5): array
{
    usort($files, fn($a, $b) => $a['timestamp'] <=> $b['timestamp']);

    $groups = [];
    $current = [];
    $expectedNextStart = null;

    foreach ($files as $file) {
        if ($current && $expectedNextStart !== null) {
            $diffMinutes = ($file['timestamp']->getTimestamp() - $expectedNextStart->getTimestamp()) / 60;
            // Разрыв между сутками (день файла отличается от дня группы) тоже
            // обязательно рвёт группу — день в любом случае не пересекаем,
            // независимо от величины разрыва.
            $sameDay = $file['timestamp']->format('Y-m-d') === $current[0]['timestamp']->format('Y-m-d');
            if (!$sameDay || $diffMinutes > $gapToleranceMinutes) {
                $groups[] = $current;
                $current = [];
            }
        }
        $current[] = $file;
        $periodMinutes = $file['period_minutes'] ?? 60; // 60 — разумное умолчание, если токен периода не распознан
        $expectedNextStart = $file['timestamp']->modify("+{$periodMinutes} minutes");
    }
    if ($current) {
        $groups[] = $current;
    }
    return $groups;
}

/**
 * Физическая склейка группы файлов (уже отсортированных хронологически —
 * rinex_group_contiguous_files() это гарантирует) в один новый файл.
 * Группа из ОДНОГО файла — это просто копирование, для единообразия
 * обработки на уровне вызывающего кода (не нужно различать "склеивать или
 * просто скопировать").
 */
function rinex_merge_group(array $localPathsInOrder, string $outputPath): bool
{
    $out = fopen($outputPath, 'wb');
    if ($out === false) {
        return false;
    }
    $headerDone = false;
    foreach ($localPathsInOrder as $path) {
        $in = fopen($path, 'rb');
        if ($in === false) {
            continue;
        }
        $skippingThisFilesHeader = $headerDone; // для первого файла группы — false, шапка пишется целиком
        while (($line = fgets($in)) !== false) {
            if ($skippingThisFilesHeader) {
                if (str_contains($line, 'END OF HEADER')) {
                    $skippingThisFilesHeader = false;
                }
                continue; // не пишем ни одной строки чужой шапки, включая саму строку-метку
            }
            fwrite($out, $line);
            if (!$headerDone && str_contains($line, 'END OF HEADER')) {
                $headerDone = true;
            }
        }
        fclose($in);
    }
    fclose($out);
    return true;
}

File: pyserver/rinex/rinex_merge.py
"""
Port of app/lib/rinex_merge.php. Stitches multiple hourly RINEX files of
one station/type into "per-day" files. Rules:
  1) Only merge files WITHOUT a gap between them -- if one file's end
     (its start + nominal duration from the filename) and the next
     file's start differ by more than gap_tolerance_minutes, that's two
     separate continuous stretches -> two separate output files, not one
     with a hole in the middle.
  2) A merge never crosses a UTC day boundary -- even back-to-back 23:00
     and 00:00-next-day files land in different days and never merge,
     mirroring how the FTP server itself is organized (one day, one
     folder).

Merging the RINEX content itself is line-based: a file has one header up
to and including "END OF HEADER", then data. For a merged file, the
header comes from the first (by time) file in the group; every other
file's header is stripped entirely, leaving only data. Works the same
for observation and navigation files -- both formats mark end-of-header
with the same "END OF HEADER" label.
"""

from datetime import timedelta


def rinex_group_contiguous_files(files: list[dict], gap_tolerance_minutes: int = 5) -> list[list[dict]]:
    """files: each a dict with 'local_path', 'timestamp' (naive datetime),
    'period_minutes' (int|None). Must already belong to one station and
    one type (MO or MN) -- that grouping happens before this call, in
    the process_rinex_requests management command."""
    files = sorted(files, key=lambda f: f["timestamp"])

    groups: list[list[dict]] = []
    current: list[dict] = []
    expected_next_start = None

    for file in files:
        if current and expected_next_start is not None:
            diff_minutes = (file["timestamp"] - expected_next_start).total_seconds() / 60
            same_day = file["timestamp"].date() == current[0]["timestamp"].date()
            if not same_day or diff_minutes > gap_tolerance_minutes:
                groups.append(current)
                current = []
        current.append(file)
        period_minutes = file.get("period_minutes") or 60  # 60 -- sane default if the period token wasn't recognized
        expected_next_start = file["timestamp"] + timedelta(minutes=period_minutes)

    if current:
        groups.append(current)
    return groups


def rinex_merge_group(local_paths_in_order: list[str], output_path: str) -> bool:
    """Physical merge of an already-chronologically-sorted group into one
    new file. A group of ONE file is just a copy, for uniform handling
    upstream (no need to distinguish "merge" from "just copy")."""
    header_done = False
    with open(output_path, "wb") as out:
        for path in local_paths_in_order:
            try:
                fh = open(path, "rb")
            except OSError:
                continue
            with fh:
                skipping_this_files_header = header_done  # False for the group's first file -- header written whole
                for line in fh:
                    if skipping_this_files_header:
                        if b"END OF HEADER" in line:
                            skipping_this_files_header = False
                        continue
                    out.write(line)
                    if not header_done and b"END OF HEADER" in line:
                        header_done = True
    return True